apps/ewallet/lib/ewallet/gates/token_gate.ex

```diff
@@ -16,69 +16,115 @@ defmodule EWallet.TokenGate do
+      decimals = attrs["subunit_to_unit"] |> :math.log10() |> trunc()
+      rootchain_identifier = BlockchainHelper.rootchain_identifier()
+      hot_wallet = BlockchainWallet.get_primary_hot_wallet(rootchain_identifier)
+
+      case BlockchainTransactionGate.deploy_erc20_token(
+             %{
+               decimals: decimals,
+               from: hot_wallet.address,
+               initial_amount: attrs["amount"],
+               locked: attrs["locked"],
+               name: attrs["name"],
+               symbol: attrs["symbol"]
+             },
+             rootchain_identifier
+           ) do
+        {:ok,
+         %{
+           contract_address: contract_address,
+           blockchain_transaction: blockchain_transaction,
+           contract_uuid: contract_uuid
+         }} ->
+          AddressTracker.register_contract_address(contract_address)
+          {:ok, put_deploy_data(attrs, blockchain_transaction, contract_address, contract_uuid)}
+
+        error ->
+          error
+      end
+    end
+  end
+
+  @spec validate_parameter_types(map()) ::
+          {:ok, map()} | {:error, :invalid_parameter, String.t()}
+  defp validate_parameter_types(
+         %{
+           "name" => name,
+           "symbol" => symbol,
+           "subunit_to_unit" => subunit_to_unit,
+           "amount" => amount,
+           "locked" => locked
+         } = parameters
+       ) do
+    with true <- is_binary(name),
+         true <- is_binary(symbol),
+         true <- is_integer(subunit_to_unit) or is_binary(subunit_to_unit),
+         true <- is_integer(amount) or is_binary(amount),
+         true <- is_boolean(locked) do
+      {:ok, parameters}
     else
-      error ->
-        error
+      _ ->
+        {:error, :invalid_parameter,
+         "`name`, `symbol`, `subunit_to_unit`, `locked` and `amount` are required when deploying an ERC20 token."}
     end
   end
 
-  def deploy_erc20(_) do
+  defp validate_parameter_types(_) do
     {:error, :invalid_parameter,
      "`name`, `symbol`, `subunit_to_unit`, `locked` and `amount` are required when deploying an ERC20 token."}
   end
 
+  @spec validate_and_normalize_attributes(map()) ::
+          {:ok, map()} | {:error, :invalid_parameter, String.t()}
+  defp validate_and_normalize_attributes(attrs) do
+    amount = normalize_value(attrs["amount"])
+    subunit_to_unit = normalize_value(attrs["subunit_to_unit"])
+
+    with true <-
+           subunit_to_unit > 0 ||
+             {:error, :invalid_parameter, "`subunit_to_unit` must be greater than 0."},
+         true <-
+           amount >= 0 ||
+             {:error, :invalid_parameter, "`amount` must be greater than or equal to 0."} do
+      normalized_attributes =
+        attrs
+        |> Map.put("amount", amount)
+        |> Map.put("subunit_to_unit", subunit_to_unit)
+
+      {:ok, normalized_attributes}
+    end
+  end
+
+  @spec normalize_value(String.t() | integer()) :: integer()
+  defp normalize_value(amount) do
```

Comment:
can be written with a guard clause at the function level
`defp normalize_value(amount) when is_binary(amount) do .....`
`defp normalize_value(amount) do .....`